Fix several bitstamp API issues

firepol opened the pull request:
Fix calls to endpoints beginning with "api/".
Allow to get all user_transactions at once (pair is optional!).
Allow to get deposit address and to withdraw BCH.
Add missing withdrawal_requests method.
Remove inexistent withdrawal_request method (withdraw exists already).
Fix transfer_sub_to_main and transfer_main_to_sub endpoints.
Various DRY-ups, clean-ups, code style improvements.

deepbrook commented on bitex/interface/bitstamp.py:
```diff
@@ -79,9 +79,8 @@ def _place_order(self, pair, price, size, side, market=None, **kwargs):
-            return self.request('%s/market/%s/' % (side, pair), authenticate=True, params=payload)
-        return self.request('%s/%s/' % (side, pair), authenticate=True, params=payload)
+        suffix = 'market/' if market else ''
+        return self.request('%s/%s%s/' % (side, suffix, pair), authenticate=True, params=payload)
```
Please use the `str.format()` method instead - I know I'm not using it either, but I too should be. It's the preferred way of formatting them.

firepol replied:
Why is it the preferred way if even you don't use it all the time? ;)

I see this notation is used in most of the interface files, that's why I kept it like that... so I'd say for the sake of consistency to leave it like that... eventually a new contributor will add new interfaces and he can use the .str() notation there. At least it should be consistent in the same file. Fair enough?

Things like this can be even noted in the contributing.md file, which I think I'll update and submit a PR for it.

deepbrook commented on bitex/interface/bitstamp.py:
```diff
@@ -109,104 +107,119 @@ def cancel_order(self, *order_ids, **kwargs):
 
     @format_with(BitstampFormattedResponse)
-    def wallet(self, *args, **kwargs):
+    def wallet(self, pair=None):
```
As with the other PRs, please see #183

firepol replied:
I realized already and re-pushed before you started the review, strange you didn't see it... in github at least if I force push the branch I used for the PR the old changes, including these comments show up as "oudated".

firepol committed:
Fix several bitstamp API issues

Fix calls to endpoints beginning with "api/".
Allow to get all user_transactions at once (pair is optional!).
Allow to get deposit address and to withdraw BCH.
Add missing withdrawal_requests method.
Remove inexistent withdrawal_request method (withdraw exists already).
Fix transfer_sub_to_main and transfer_main_to_sub endpoints.
Various DRY-ups, clean-ups, code style improvements.

firepol commented:
@nlsdfnbch : based on your feedback on my other PRs I kept the args/kwargs in the standardized methods, but removed it when not needed, in the exchange specific methods.

Also, the withdraw/deposit methods were not supporting all cryptos, e.g. BCH was missing. I initially did some logic to check the valid ones, but based on your feedback on cryptopia it seems better not to bother, if the user tries to withdraw an invalid currency, the API will fail and an exception will be thrown.

I noticed that bitcoin and ripple used a different notation (currency entire name, not symbol), but the new ones use the symbol, so like this if bitstamp adds new coins, it should work also for the future ones, unless they become "creative" and change naming convention and become inconsistent.

deepbrook commented on bitex/interface/bitstamp.py:
```diff
@@ -77,136 +77,129 @@ def bid(self, pair, price, size, *args, market=False, **kwargs):
-        if pair:
-            return self.request('open_orders/%s/' % pair, authenticate=True, params=kwargs)
-        return self.request('open_orders/all/', authenticate=True, params=kwargs)
+        suffix = pair if pair else 'all'
```
Well, while you're at it might as well do `pair or 'all'` ;)

firepol replied:
good one, thx I think I did something similar in another file (cryptopia or bittrex) but I guess we'll live with it. Done.

deepbrook commented on bitex/interface/bitstamp.py:
```diff
@@ -77,136 +77,129 @@ def bid(self, pair, price, size, *args, market=False, **kwargs):
 
     @format_with(BitstampFormattedResponse)
-    def wallet(self, *args, **kwargs):
+    def wallet(self, pair=None, *args, **kwargs):
```
please keep `args` and `kwargs` in order

firepol replied:
done

firepol commented:
@nlsdfnbch I think this PR is ready for merging, I fixed it based on your feedback. Anything else blocking the merge?

deepbrook commented:
There are a few minor things, but I haven't had the time to write them up. I'll do so later today

deepbrook commented on bitex/interface/bitstamp.py:
```diff
@@ -77,136 +77,129 @@ def bid(self, pair, price, size, *args, market=False, **kwargs):
-
-            return self.request('balance/%s/' % pair, authenticate=True, params=kwargs)
-        return self.request('balance/', authenticate=True, params=kwargs)
+                pair = pair
```
this is redundant. Instead, log a message on debug level along the lines of "pair is not a Formatter object, assuming string.." for debug purposes.

firepol replied:
I found a better way: `pass`

deepbrook commented on bitex/interface/bitstamp.py:
```diff
@@ -77,136 +77,129 @@ def bid(self, pair, price, size, *args, market=False, **kwargs):
 
     @check_and_format_pair
-    def hourly_ticker(self, pair, **kwargs):
+    def hourly_ticker(self, pair):
```
if it's optional, make it a kwarg instead.

firepol replied:
I know it's optional, I think it's bad design of the bitstamp API. In all the other methods, because they are basic methods, `pair` was added to the method signature, the only missing method to have everything consistent at least in bitstamp to bitex is this method, thus I'd suggest to keep pair as parameter in the signature. Else in all the other methods you do like this:
```
api = bitex.Bitstamp()
btcusd_order_book = api.order_book('btcusd')
btcusd_transactions = api.trades('btcusd')
btc_usd_open_orders = api.open_orders('btcusd')
btcusd_ticker = api.ticker('btcusd')
```

And now the hourly ticker:

```
btcusd_hourly_ticker = api.ticker('btcusd')
```

With kwargs:

```
btcusd_ticker = api.ticker(**{'pair': 'btcusd'})
```

A bit ugly & inconsistent, don't you think?

But it should not contain "if pair", I should rather fix the endpoint to include it directly, like this:

```
    @check_and_format_pair
    def hourly_ticker(self, pair):
        """Return the hourly ticker for the given pair."""
        return self.request('ticker_hour/' + pair)
```

Agree?

firepol replied:
Actually, checking again the official bitstamp API, v2 **requires** the pair parameter in this method.

https://www.bitstamp.net/api/#ticker-hour

deepbrook commented:
Actually, python is smart about  its parameters - if you pass pair as a positional argument, but no positional argument is defined in the function signature, it'll automatically assign it to the first explicitly defined kwarg. So, despite having it defined as a keyword argument, you can still call it like so:
```python
b.hourly_ticker('btcusd')
```

On 8 May 2018 19:42:35 CEST, firepol <notifications@github.com> wrote:
>firepol commented on this pull request.
>
>
>
>>              except AttributeError:
>-                pair = kwargs['pair']
>-
>-            return self.request('balance/%s/' % pair,
>authenticate=True, params=kwargs)
>-        return self.request('balance/', authenticate=True,
>params=kwargs)
>+                pair = pair
>
>I found a better way: `pass`
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/Crypto-toolbox/bitex/pull/184#discussion_r186810257

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

firepol commented:
Hi,

I tried it using this mwethod with kwargs:

def hourly_ticker(self, **kwargs):
    """Return the hourly ticker for the given pair."""
    pair = kwargs.get('pair') or ''
    return self.request('ticker_hour/' + pair)

In python 3.6 this doesn't work for me:

import bitex
bitstamp = bitex.Bitstamp()
hourly = bitstamp.hourly_ticker('xrpusd').json()
print(hourly)

I get this error:

    hourly = bitstamp.hourly_ticker('xrpusd').json()
TypeError: hourly_ticker() takes 1 positional argument but 2 were given

Also, the bitstamp v2 API actually requires the pair, so I really don't see
why not just putting it in the method...
Bitstamp v1 api seems it was done when they were actually offering only
btcusd (thats' why lots of old methods don't need a parameter and actually
offer btcusd as default), super old now, nobody should use it...


On Tue, May 8, 2018 at 8:00 PM, Nils Diefenbach <notifications@github.com>
wrote:

> Actually, python is smart about its parameters - if you pass pair as a
> positional argument, but no positional argument is defined in the function
> signature, it'll automatically assign it to the first explicitly defined
> kwarg. So, despite having it defined as a keyword argument, you can still
> call it like so:
> ```python
> b.hourly_ticker('btcusd')
> ```
>
> On 8 May 2018 19:42:35 CEST, firepol <notifications@github.com> wrote:
> >firepol commented on this pull request.
> >
> >
> >
> >> except AttributeError:
> >- pair = kwargs['pair']
> >-
> >- return self.request('balance/%s/' % pair,
> >authenticate=True, params=kwargs)
> >- return self.request('balance/', authenticate=True,
> >params=kwargs)
> >+ pair = pair
> >
> >I found a better way: `pass`
> >
> >--
> >You are receiving this because you were mentioned.
> >Reply to this email directly or view it on GitHub:
> >https://github.com/Crypto-toolbox/bitex/pull/184#discussion_r186810257
>
> --
> Sent from my Android device with K-9 Mail. Please excuse my brevity.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Crypto-toolbox/bitex/pull/184#issuecomment-387489654>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABn7PoTzK9NZ19lE7hz7r0LgSxHSJRgtks5twd1IgaJpZM4TgsLT>
> .
>

deepbrook commented:
You misunderstood me. Try this:
```
def foo(bar=None, **kwargs):
    print(bar)

foo("cheese balls")
```

On 8 May 2018 22:00:34 CEST, firepol <notifications@github.com> wrote:
>Hi,
>
>I tried it using this mwethod with kwargs:
>
>def hourly_ticker(self, **kwargs):
>    """Return the hourly ticker for the given pair."""
>    pair = kwargs.get('pair') or ''
>    return self.request('ticker_hour/' + pair)
>
>In python 3.6 this doesn't work for me:
>
>import bitex
>bitstamp = bitex.Bitstamp()
>hourly = bitstamp.hourly_ticker('xrpusd').json()
>print(hourly)
>
>I get this error:
>
>    hourly = bitstamp.hourly_ticker('xrpusd').json()
>TypeError: hourly_ticker() takes 1 positional argument but 2 were given
>
>Also, the bitstamp v2 API actually requires the pair, so I really don't
>see
>why not just putting it in the method...
>Bitstamp v1 api seems it was done when they were actually offering only
>btcusd (thats' why lots of old methods don't need a parameter and
>actually
>offer btcusd as default), super old now, nobody should use it...
>
>
>On Tue, May 8, 2018 at 8:00 PM, Nils Diefenbach
><notifications@github.com>
>wrote:
>
>> Actually, python is smart about its parameters - if you pass pair as
>a
>> positional argument, but no positional argument is defined in the
>function
>> signature, it'll automatically assign it to the first explicitly
>defined
>> kwarg. So, despite having it defined as a keyword argument, you can
>still
>> call it like so:
>> ```python
>> b.hourly_ticker('btcusd')
>> ```
>>
>> On 8 May 2018 19:42:35 CEST, firepol <notifications@github.com>
>wrote:
>> >firepol commented on this pull request.
>> >
>> >
>> >
>> >> except AttributeError:
>> >- pair = kwargs['pair']
>> >-
>> >- return self.request('balance/%s/' % pair,
>> >authenticate=True, params=kwargs)
>> >- return self.request('balance/', authenticate=True,
>> >params=kwargs)
>> >+ pair = pair
>> >
>> >I found a better way: `pass`
>> >
>> >--
>> >You are receiving this because you were mentioned.
>> >Reply to this email directly or view it on GitHub:
>>
>>https://github.com/Crypto-toolbox/bitex/pull/184#discussion_r186810257
>>
>> --
>> Sent from my Android device with K-9 Mail. Please excuse my brevity.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>>
><https://github.com/Crypto-toolbox/bitex/pull/184#issuecomment-387489654>,
>> or mute the thread
>>
><https://github.com/notifications/unsubscribe-auth/ABn7PoTzK9NZ19lE7hz7r0LgSxHSJRgtks5twd1IgaJpZM4TgsLT>
>> .
>>
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/Crypto-toolbox/bitex/pull/184#issuecomment-387524298

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

firepol commented:
Aaaah ok I get it now ;)

But I still don't understand what would be the code change to do here.

Let's have a look at the old method:

    @check_and_format_pair
    def hourly_ticker(self, pair, **kwargs):
        """Return the hourly ticker for the given pair."""
        if pair:
            return self.request('ticker_hour/%s/' % pair, params=kwargs)
        return self.request('api/ticker_hour/')

New method:

    @check_and_format_pair
    def hourly_ticker(self, pair):
        """Return the hourly ticker for the given pair."""
        return self.request('ticker_hour/' + pair)

I removed kwargs, because this is not a standardized method and also
because no additional parameters are needed here.

The old method supports a call without "pair", but it's wrong: the method
signature should have been pair=None. Anyway, even with fixed signature: If
no pair is given, by default it'll return the hourly ticker for btcusd
using the old API call. Not really needed on my opinion (because it makes
more sense to pass a pair, not just to expect a default, even if the
bitstamp api allows that), thus I cleaned up the code like that.

I removed kwargs also in: eur_usd_conversion_rate(self)

Hope it works for you?


On Tue, May 8, 2018 at 10:50 PM, Nils Diefenbach <notifications@github.com>
wrote:

> You misunderstood me. Try this:
> ```
> def foo(bar=None, **kwargs):
> print(bar)
>
> foo("cheese balls")
> ```
>
>
> On 8 May 2018 22:00:34 CEST, firepol <notifications@github.com> wrote:
> >Hi,
> >
> >I tried it using this mwethod with kwargs:
> >
> >def hourly_ticker(self, **kwargs):
> > """Return the hourly ticker for the given pair."""
> > pair = kwargs.get('pair') or ''
> > return self.request('ticker_hour/' + pair)
> >
> >In python 3.6 this doesn't work for me:
> >
> >import bitex
> >bitstamp = bitex.Bitstamp()
> >hourly = bitstamp.hourly_ticker('xrpusd').json()
> >print(hourly)
> >
> >I get this error:
> >
> > hourly = bitstamp.hourly_ticker('xrpusd').json()
> >TypeError: hourly_ticker() takes 1 positional argument but 2 were given
> >
> >Also, the bitstamp v2 API actually requires the pair, so I really don't
> >see
> >why not just putting it in the method...
> >Bitstamp v1 api seems it was done when they were actually offering only
> >btcusd (thats' why lots of old methods don't need a parameter and
> >actually
> >offer btcusd as default), super old now, nobody should use it...
> >
> >
> >On Tue, May 8, 2018 at 8:00 PM, Nils Diefenbach
> ><notifications@github.com>
> >wrote:
> >
> >> Actually, python is smart about its parameters - if you pass pair as
> >a
> >> positional argument, but no positional argument is defined in the
> >function
> >> signature, it'll automatically assign it to the first explicitly
> >defined
> >> kwarg. So, despite having it defined as a keyword argument, you can
> >still
> >> call it like so:
> >> ```python
> >> b.hourly_ticker('btcusd')
> >> ```
> >>
> >> On 8 May 2018 19:42:35 CEST, firepol <notifications@github.com>
> >wrote:
> >> >firepol commented on this pull request.
> >> >
> >> >
> >> >
> >> >> except AttributeError:
> >> >- pair = kwargs['pair']
> >> >-
> >> >- return self.request('balance/%s/' % pair,
> >> >authenticate=True, params=kwargs)
> >> >- return self.request('balance/', authenticate=True,
> >> >params=kwargs)
> >> >+ pair = pair
> >> >
> >> >I found a better way: `pass`
> >> >
> >> >--
> >> >You are receiving this because you were mentioned.
> >> >Reply to this email directly or view it on GitHub:
> >>
> >>https://github.com/Crypto-toolbox/bitex/pull/184#discussion_r186810257
> >>
> >> --
> >> Sent from my Android device with K-9 Mail. Please excuse my brevity.
> >>
> >> —
> >> You are receiving this because you authored the thread.
> >> Reply to this email directly, view it on GitHub
> >>
> ><https://github.com/Crypto-toolbox/bitex/pull/184#issuecomment-387489654
> >,
> >> or mute the thread
> >>
> ><https://github.com/notifications/unsubscribe-auth/
> ABn7PoTzK9NZ19lE7hz7r0LgSxHSJRgtks5twd1IgaJpZM4TgsLT>
> >> .
> >>
> >
> >
> >--
> >You are receiving this because you were mentioned.
> >Reply to this email directly or view it on GitHub:
> >https://github.com/Crypto-toolbox/bitex/pull/184#issuecomment-387524298
>
>
> --
> Sent from my Android device with K-9 Mail. Please excuse my brevity.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Crypto-toolbox/bitex/pull/184#issuecomment-387538037>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABn7PloeINafDw8zWvNpPrH7SwHR4FeFks5twgT-gaJpZM4TgsLT>
> .
>
